x/ibctransfer/whitelisting.go

```diff
@@ -5,48 +5,149 @@ import (
 				sdk.NewAttribute(transfertypes.AttributeKeyAmount, fmt.Sprintf("%d", data.Amount)),
 				sdk.NewAttribute(transfertypes.AttributeKeyAckSuccess, fmt.Sprintf("%t", false)),
 			),
 		)
-
 		return &sdk.Result{
 			Events: ctx.EventManager().Events().ToABCIEvents(),
 		}, acknowledgement.GetBytes(), nil
 	}
+	// get result of transfer receive
+	acknowledgement := channeltypes.NewResultAcknowledgement([]byte{byte(1)})
+	err := sdkTransferKeeper.OnRecvPacket(ctx, packet, data)
+	if err != nil {
+		acknowledgement = channeltypes.NewErrorAcknowledgement(err.Error())
+	}
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			transfertypes.EventTypePacket,
+			sdk.NewAttribute(sdk.AttributeKeyModule, transfertypes.ModuleName),
+			sdk.NewAttribute(transfertypes.AttributeKeyReceiver, data.Receiver),
+			sdk.NewAttribute(transfertypes.AttributeKeyDenom, data.Denom),
+			sdk.NewAttribute(transfertypes.AttributeKeyAmount, fmt.Sprintf("%d", data.Amount)),
+			sdk.NewAttribute(transfertypes.AttributeKeyAckSuccess, fmt.Sprintf("%t", err == nil)),
+		),
+	)
+	// NOTE: acknowledgement will be written synchronously during IBC handler execution.
+	recvResult := &sdk.Result{
+		Events: ctx.EventManager().Events().ToABCIEvents(),
+	}
+	resBytes := acknowledgement.GetBytes()
+
+	// if no error and packet is returning and needs conversion: convert
+	if err != nil && IsRecvPacketReturning(packet, data) && shouldConvertDecimals(ctx, whitelistKeeper, packet, data) {
+		ibcToken, convToken := convertDecimals(ctx, whitelistKeeper, packet, data)
+		recvResult, err = sendConvertRecvDenom(ctx, ibcToken, convToken, bankKeeper, data)
+	}
+	// otherwise return
+	return recvResult, resBytes, err
+}
+
+func shouldConvertDecimals(
+	ctx sdk.Context,
+	whitelistKeeper tokenregistrytypes.Keeper,
+	packet channeltypes.Packet,
+	data transfertypes.FungibleTokenPacketData,
+) bool {
+	// get token registry entry for received denom
+	denom := GetMintedDenomFromPacket(packet, data)
+	registryEntry := whitelistKeeper.GetRegistryEntry(ctx, denom)
+	// if decimals are greater than ibc decimals, we need to increase precision to convert them
+	return registryEntry.IbcDenom != "" && registryEntry.Decimals > registryEntry.IbcDecimals
+}
+
+func convertDecimals(
+	ctx sdk.Context,
+	whitelistKeeper tokenregistrytypes.Keeper,
+	packet channeltypes.Packet,
+	data transfertypes.FungibleTokenPacketData,
+) (sdk.Coin, sdk.Coin) {
+	// get token registry entry for received denom
+	denom := GetMintedDenomFromPacket(packet, data)
+	registryEntry := whitelistKeeper.GetRegistryEntry(ctx, denom)
+	// get the token amount from the packet data
+	decAmount := sdk.NewDecFromInt(sdk.NewIntFromUint64(data.Amount))
+	// calculate the conversion difference and increase precision
+	po := registryEntry.Decimals - registryEntry.IbcDecimals
+	convAmountDec := IncreasePrecision(decAmount, po)
+	convAmount := sdk.NewIntFromBigInt(convAmountDec.RoundInt().BigInt())
```

Comment:
Should be changed in case we decided to move to truncate .

Reply:
I think he's suggesting that we use Truncate() instead of RoundInt(). 

This makes me think of another issue, if a user sends 1000000000010000000 rowan we:
- calculate convToken to be 10000000000 urowan
- send 1000000000010000000 rowan from the user to the module
- mint 10000000000 urowan
- send 10000000000 urowan via ibc
Then if the user sends the same tokens back we:
- calculate convToken to be 1000000000000000000 rowan
- send 10000000000 urowan from user to the module
- send 1000000000000000000 rowan from the module to the user
- burn 10000000000 urowan
So the user lost 10000000 rowan. I think we need to reduce precision on the send, then increase precision on that amount 
and use that for the amount that's deducted from the user.

Does this make sense to you guys?
@timlind @Jedi2002 @iverc

Reply:
Yeah that makes sense , 
I was actually suggesting that we always round down ,so that the network never looses an tokens as an additional safety .

Reply:
Yeah I agree we should truncate. And on top of that we should determine the amount of tokens to deduct based on the truncated amount.

Reply:
👍🏻

Reply:
yeah, rounding down was discussed a couple of times, so instead of bankers rounding we always round down so we never send more that needed / available. Fixing that logic would require updating test tables since expected values are calculated based on bankers rounding logic